service/controller/resource/nodepool/template/template.go

```diff
@@ -23,3 +25,65 @@ func GetARMTemplate() (map[string]interface{}, error) {
 	}
 	return contents, microerror.Mask(err)
 }
+
+func NewDeployment(templateParams Parameters) (azureresource.Deployment, error) {
+	armTemplate, err := GetARMTemplate()
+	if err != nil {
+		return azureresource.Deployment{}, microerror.Mask(err)
+	}
+
+	return azureresource.Deployment{
+		Properties: &azureresource.DeploymentProperties{
+			Mode:       azureresource.Incremental,
+			Parameters: templateParams.ToDeployParams(),
+			Template:   armTemplate,
+		},
+	}, nil
+}
+
+// IsOutOfDate returns whether or not two Azure ARM Deployments are equal.
+func IsOutOfDate(currentDeployment azureresource.DeploymentExtended, desiredDeployment azureresource.Deployment) (bool, error) {
+	if currentDeployment.IsHTTPStatus(404) {
+		return true, nil
+	}
+
+	currentParameters, err := NewFromExtendedDeployment(currentDeployment)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	desiredParameters, err := NewFromDeployment(desiredDeployment)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	return !reflect.DeepEqual(currentParameters, desiredParameters), nil
+}
+
+// NeedToRolloutNodes tells whether or not we need to replace the existing VMs.
+// We only don't roll out nodes when we change the scaling parameters.
+func NeedToRolloutNodes(currentDeployment azureresource.DeploymentExtended, desiredDeployment azureresource.Deployment) (bool, error) {
+	currentParameters, err := NewFromExtendedDeployment(currentDeployment)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	desiredParameters, err := NewFromDeployment(desiredDeployment)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	rolloutnodes := currentParameters.AzureOperatorVersion != desiredParameters.AzureOperatorVersion ||
+		currentParameters.ClusterID != desiredParameters.ClusterID ||
```

Comment:
Would it make any sense to inject logger to this function and log changed parameters by wrapping these comparisons with something like:

```go
func logChange(ctx context.Context, logger micrologger.Logger, annotation string, v bool) {
	if !v {
		logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("change detected for %q", annotation))
	}
}
```

Then each comparison line would be like:
```go
		logChange(ctx, logger, "ClusterID", (currentParameters.ClusterID != desiredParameters.ClusterID)) ||
```

Reply:
not sure I understand how the code would look like for `rolloutnodes`?

Reply:
I guess `logChange` needs to return `v`?

Reply:
Oh, yes, somehow missed that when writing code in separate comment :slightly_smiling_face: